File: day01/taxi2.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.24.0;

my %dirv = (N => [0,1],
            S => [0,-1],
            E => [1,0],
            W => [-1,0]
           );

my %r = (N => 'E',
         E => 'S',
         S => 'W',
         W => 'N'
        );

my %l = (N => 'W',
         E => 'N',
         S => 'E',
         W => 'S'
        );

my %visited;

my @pos = (0, 0);
my $dir = 'N';
my $done = 0;

while (<>) {
    chomp;
    my @moves = split /, /;
    for my $move (@moves) {
        my $turn = substr($move, 0, 1);
        my $n = substr($move, 1);

        if ($turn eq 'R') {
            $dir = $r{$dir};
        } else {
            $dir = $l{$dir};
        }
        my $dirv = $dirv{$dir};

        for my $j (0..$n-1) {
            $pos[0] += $dirv->[0];
            $pos[1] += $dirv->[1];

            if (++$visited{"@pos"} > 1) {
                say "visited (@pos) twice!";
                $done = 1;
                last;
            }
        }
        last if $done;
    }
}

if ($done) {
    say "@pos";
    say abs($pos[0]) + abs($pos[1]);
} else {
    say "Santa didn't visit any location twice.";
}

File: day01/taxi2.py
#!/usr/bin/env python
from sys import argv
from numpy import *

dirv = {}
dirv['N'] = array([0, 1])
dirv['S'] = array([0,-1])
dirv['E'] = array([1, 0])
dirv['W'] = array([-1,0])

r = {}
r['N'] = 'E'
r['E'] = 'S'
r['S'] = 'W'
r['W'] = 'N'

l = {}
l['N'] = 'W'
l['E'] = 'N'
l['S'] = 'E'
l['W'] = 'S'

visited = {}

pos = array([0, 0])
dir = 'N'
done = False

script, filename = argv
with open(filename) as f :
    for line in f :
        line = line.rstrip("\n")
        moves = line.split(", ")
        for move in moves :
            turn = move[:1]
            n = int(move[1:])

            if turn == "R" :
                dir = r[dir]
            else :
                dir = l[dir]

            v = dirv[dir]
            for i in range(0,n) :
                pos = pos + v
                k = "%d %d" % (pos[0], pos[1])
                if k in visited :
                    print "visited %s twice!" % pos
                    done = True
                    break
                else :
                    visited[k] = 1
                
            if done :
                break

if done :
    print pos
    print abs(pos[0]) + abs(pos[1])
else :
    print "Santa didn't visit any location twice"
    
